### PR TITLE
Feature/markup frontstage

### PR DESCRIPTION
**Pull request by archana-maharjan**
This PR contains the markup frontstage that provides markup frontstage provider to add markup.

### CODE REVIEW

**Comment by archana-maharjan** [comment]
I had issue with setting up the test, I need help on resolving following error

>  Test suite failed to run
>     /Users/archanamaharjan/Desktop/dev/viewer-components-react/common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/lib/ui-core/colorthemes.scss:5
>     @import "./colorvariables.scss";
>     ^
>     SyntaxError: Invalid or unexpected token
>       at ScriptTransformer._transformAndBuildScript (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
>       at ScriptTransformer.transform (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:579:25)
>       at Object.<anonymous> (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/src/ui-core/UiCore.ts:10:1)

**Comment by archana-maharjan** [comment]
> I had issue with setting up the test, I need help on resolving following error
> 
> > Test suite failed to run
> > /Users/archanamaharjan/Desktop/dev/viewer-components-react/common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/lib/ui-core/colorthemes.scss:5
> > @import "./colorvariables.scss";
> > ^
> > SyntaxError: Invalid or unexpected token
> > at ScriptTransformer._transformAndBuildScript (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
> > at ScriptTransformer.transform (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:579:25)
> > at Object. (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/src/ui-core/UiCore.ts:10:1)

**Comment by archana-maharjan** [comment]
@Carl-Hinkle  Thank you very much for helping me resolve import issue in test suite.

**Comment by archana-maharjan** [comment]
> I had issue with setting up the test, I need help on resolving following error
> 
> > Test suite failed to run
> > /Users/archanamaharjan/Desktop/dev/viewer-components-react/common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/lib/ui-core/colorthemes.scss:5
> > @import "./colorvariables.scss";
> > ^
> > SyntaxError: Invalid or unexpected token
> > at ScriptTransformer._transformAndBuildScript (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:537:17)
> > at ScriptTransformer.transform (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@jest/transform/24.9.0/node_modules/@jest/transform/build/ScriptTransformer.js:579:25)
> > at Object. (../../common/temp/node_modules/.pnpm/registry.npmjs.org/@bentley/ui-core/2.2.0_22a535e34cdd435cd9ce754bca6cd5d1/node_modules/@bentley/ui-core/src/ui-core/UiCore.ts:10:1)

@Carl-Hinkle Thank you very much for helping me resolve import issue in test suite.

File: .gitignore [hunk 1]
```diff
@@ -36,3 +36,7 @@ common/temp/**
 package-deps.json
 *.build.log
 *.build.error.log
+
+
+#igoner DS store
+.DS_Store
```
Comment (calebmshafer) [inline .gitignore:42]: Looks like one of these was added under `packages/.DS_store`, can you remove it?

File: .gitignore [hunk 2]
```diff
@@ -36,3 +36,7 @@ common/temp/**
 package-deps.json
 *.build.log
 *.build.error.log
+
```
Comment (calebmshafer) [inline .gitignore:39]: nit: extra line

File: packages/markup-frontstage/README.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
+# @bentley\markup-frontstage-react
+
+Copyright © Bentley Systems, Incorporated. All rights reserved.
+
+The markup-frontstage-react provides Markup Frontstage for adding, editing, deleting markup elements in iModel.js application.
```
Comment (calebmshafer) [inline packages/markup-frontstage/README.md:5]: Missing a determiner (I think that's what they're called) here, I'd go with something like "provides a Markup Frontstage" or "provides the Markup Frontstage".

File: packages/markup-frontstage/README.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+## MarkupFrontstageProvider
+
+FrontstageProvider object that requires ViewState, IModelConnection, Editable boolean flag whether or not the markup elements are editable, ConfigurableUiControlContructor, Svg string as optional, ViewElementDictionary with 9 zone layouts to be added in the frontstage as optional, EmphasizeElementsProps as optional to instantiate. It's onAddMarkupEvent allows to raise event on adding markup, onstopMarkupEvent allows to raise event on stopping markup. Additionally there's onCloseAsync and onSaveAsync properties to be called to close the Frontstage and Save the Markup data respectively.
+
+## MarkupSettingsPanel
+
+It is the flyover component used to set colors, width, opacity of the markup element by MarkupFronstage.
```
Comment (calebmshafer) [inline packages/markup-frontstage/README.md:13]: misspelling here, MarkupFrontstage

File: packages/markup-frontstage/src/components/toolbar/MarkupToolWidget.module.scss [hunk 5]
```diff
@@ -0,0 +1,7 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+.verticalToolbar {
```
[the remark this answers is not in the document]
Comment (calebmshafer) [reply packages/markup-frontstage/src/components/toolbar/MarkupToolWidget.module.scss:5]: talked to @archana-maharjan and this is not needed since it's a module.scss file and should use the tag `<FileName>_verticalToolbar_<4GuidLetters>`.

File: .gitignore (hunk 1 above)
Comment (calebmshafer) [reply .gitignore]: @archana-maharjan here I was saying that the actual `.DS_Store` files have been added in this PR. You can leave the entry for it in the `.gitignore`.

File: packages/markup-frontstage/.gitignore [hunk 6]
```diff
@@ -0,0 +1,42 @@
+# Logs
+*.log
```
[the remark this answers is not in the document]
Comment (archana-maharjan) [reply packages/markup-frontstage/.gitignore]: No sir, added that just for DS store which didn't worked from root because I had spelling mistake, Carl found it and changed it on both but forgot to remove from makrup-frontstage, for now it's removed please :)

File: .gitignore (hunk 1 above)
Comment (archana-maharjan) [reply .gitignore:41]: Lol, sorry I misread your statement. Both DS_store removed please!

File: packages/markup-frontstage/README.md (hunk 4 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/README.md]: Added space in between the words!

Comment (archana-maharjan) [reply packages/markup-frontstage/README.md]: I thought it was just the space between, now I see Frontstage spelling was wrong, corrected please :)

File: packages/markup-frontstage/README.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# @bentley\markup-frontstage-react
```
Comment (aruniverse) [inline packages/markup-frontstage/README.md:1]: should be @bentley/markup-frontstage-react

File: packages/markup-frontstage/declarations.d.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+/*---------------------------------------------------------------------------------------------
```
Comment (aruniverse) [inline packages/markup-frontstage/declarations.d.ts:1]: was this file actually needed? I recall you adding this to try and get tests working when importing scss files in components

File: packages/markup-frontstage/jest.config.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+module.exports = {
```
Comment (aruniverse) [inline packages/markup-frontstage/jest.config.js:5]: any reason the jest config cant be moved back into the package.json? or move the rest of the jest config in here?

File: packages/markup-frontstage/package.json [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  "description": "Markup Frontstage React",
+  "license": "MIT",
+  "main": "lib/markup-frontstage-react.js",
+  "typings": "lib/markup-frontstage-react",
+  "scripts": {
+    "copy:assets": "cpx \"./src/**/*.scss\" ./lib && cpx \"./public/**/*\" ./lib/public",
+    "pseudolocalize": "betools pseudolocalize --englishDir ./public/locales/en --out ./lib/public/locales/en-PSEUDO",
+    "build": " tsc 1>&2 && npm run copy:assets && npm run pseudolocalize && npm run copy-css",
+    "build-prod": " tsc -p tsconfig.json 1>&2 && npm run copy:assets && npm run copy-css",
+    "clean": "rimraf lib build",
```
Comment (aruniverse) [inline packages/markup-frontstage/package.json:13]: the build dir youre trying to delete should never exist anyway, you can remove that

File: packages/markup-frontstage/package.json [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    "@bentley/build-tools": "^2.0.0",
+    "@bentley/imodeljs-common": "^2.0.0",
+    "@bentley/imodeljs-frontend": "^2.0.0",
+    "@bentley/imodeljs-i18n": "^2.0.0",
+    "@bentley/imodeljs-markup": "^2.0.0",
+    "@bentley/presentation-frontend": "^2.0.0",
+    "@bentley/ui-abstract": "^2.0.0",
+    "@bentley/ui-components": "^2.0.0",
+    "@bentley/ui-core": "^2.0.0",
+    "@bentley/ui-framework": "^2.0.0",
```
Comment (aruniverse) [inline packages/markup-frontstage/package.json:53]: core bentley imjs deps shouldnt be using `^` and should be pinned

File: packages/markup-frontstage/package.json [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+    "@types/i18next": "^8.4.2",
+    "@types/jest": "^24.0.15",
+    "@types/react": "16.9.23",
+    "@types/react-dom": "16.9.5",
+    "@types/redux-mock-store": "^1.0.2",
+    "babel-jest": "^24.9.0",
+    "babel-plugin-import-remove-resource-query": "^1.0.0",
+    "concurrently": "^4.1.0",
+    "copyfiles": "^2.1.0",
+    "cpx": "^1.5.0",
```
Comment (aruniverse) [inline packages/markup-frontstage/package.json:65]: i recommend either using cpx or copyfiles, really no need for both. scripts should be updated accordingly

File: packages/markup-frontstage/src/MarkupFrontstage.ts [hunk 13]
```diff
@@ -0,0 +1,77 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @module Common */
```
Comment (aruniverse) [inline packages/markup-frontstage/src/MarkupFrontstage.ts:5]: do you need this line? `/** @module Common */`

File: packages/markup-frontstage/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+    "jsx": "react",
+    "declaration": true,
+    "sourceMap": true,
+    "inlineSources": true,
+    "outDir": "./lib",
+    "downlevelIteration": true,
+    "allowJs": true
+  },
+  "include": ["./src/**/*.ts*", "declarations.d.ts"],
+  "exclude": ["lib", "build", "node_modules"]
```
Comment (aruniverse) [inline packages/markup-frontstage/tsconfig.json:30]: no build dir

File: packages/markup-frontstage/src/MarkupFrontstage.ts [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+      );
+    }
+    return MarkupFrontstage._i18n;
+  }
+
+  /**
+   * The internationazlization service namespace.
+   */
+  public static get i18nNamespace(): string {
+    return "MakrupFrontstage";
```
Comment (JoeZman) [inline packages/markup-frontstage/src/MarkupFrontstage.ts:53]: misspelling... MakrupFrontstage?

File: packages/markup-frontstage/jest.config.js (hunk 9 above)
Comment (Carl-Hinkle) [reply packages/markup-frontstage/jest.config.js:5]: Yes.  The reason that the tests were failing is because CRA only supports a small subset of jest configuration options in the package.json file.  If you include jest.config.js, you can include the `moduleNameMapper` property, which is necessary in order to use @import in the ui-core scss files.  Previously `moduleNameMapper` was in the package.json file, but it didn't do anything at all.

File: packages/markup-frontstage/src/util/MarkupTypes.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+          widgets={[
+            <Widget
+              key={MarkupFrontstageConstants.TOOL_SETTINGS_WIDGET_KEY}
+              isToolSettings={true}
+            />,
+          ]}
+        />
+      )
+    }
+ */
```
Comment (JoeZman) [inline packages/markup-frontstage/src/util/MarkupTypes.ts:36]: Are you leaving this commented out code snippet for a reason?

File: packages/markup-frontstage/declarations.d.ts (hunk 8 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/declarations.d.ts:1]: It is needed, I wonder why would you want to remove it?

File: packages/markup-frontstage/src/util/MarkupTypes.ts (hunk 16 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/src/util/MarkupTypes.ts:36]: Yes sir, it's the example for what is expected in ViewElementDictionary :)

File: packages/markup-frontstage/README.md (hunk 7 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/README.md]: thanks, corrected please :)

File: packages/markup-frontstage/src/MarkupFrontstage.ts (hunk 13 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/src/MarkupFrontstage.ts]: removed please :)

File: packages/markup-frontstage/src/MarkupFrontstage.ts (hunk 15 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/src/MarkupFrontstage.ts]: Thanks, corrected please :)

File: packages/markup-frontstage/src/components/ui/MarkupFrontstageProvider.tsx [hunk 17]
```diff
@@ -0,0 +1,349 @@
...
+    if (newViewPort) {
+      const oldViewState = newViewPort.view.clone();
+      const oldEmphasizedElementsProps = this._emphasizedElementsProps;
+      FrontstageManager.onFrontstageReadyEvent.addOnce(() => {
+        const selectedView = IModelApp.viewManager.selectedView;
+        if (selectedView && oldViewState) {
+          // change viewflags here if needed, should be before following.
+          selectedView.changeView(oldViewState);
+        }
+        // As per Diego, emphasized elements are not saved in view state, which we have to preserve manually as follow
```
Comment (aruniverse) [inline packages/markup-frontstage/src/components/ui/MarkupFrontstageProvider.tsx:152]: Should probably not reference anyone who is not a code owner in coments

File: packages/markup-frontstage/package.json (hunk 12 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/package.json:63]: completely forgot to remove it while I and Jason were trying to fix css problem, thanks, removed now!

File: packages/markup-frontstage/src/components/ui/MarkupFrontstageProvider.tsx (hunk 17 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/src/components/ui/MarkupFrontstageProvider.tsx]: Sad to remove his, name wish I could refer him as IModelTeam or something that could go cool with :) well, now we know who to bug :)

File: packages/markup-frontstage/package.json (hunk 10 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/package.json]: done

File: packages/markup-frontstage/tsconfig.json (hunk 14 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/tsconfig.json]: done

File: packages/markup-frontstage/package.json [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+  "version": "1.0.0",
+  "description": "Markup Frontstage React",
+  "license": "MIT",
+  "main": "lib/markup-frontstage-react.js",
+  "typings": "lib/markup-frontstage-react",
+  "scripts": {
+    "copy:assets": "cpx \"./src/**/*.scss\" ./lib && cpx \"./public/**/*\" ./lib/public",
+    "pseudolocalize": "betools pseudolocalize --englishDir ./public/locales/en --out ./lib/public/locales/en-PSEUDO",
+    "build": " tsc 1>&2 && npm run copy:assets && npm run pseudolocalize",
+    "build-prod": " tsc -p tsconfig.json 1>&2 && npm run copy:assets",
```
Comment (calebmshafer) [inline packages/markup-frontstage/package.json:12]: This script will never be run, can probably be removed.

File: packages/markup-frontstage/package.json [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    "clean": "rimraf lib",
+    "lint": "tslint -p . 1>&2",
```
Comment (calebmshafer) [inline packages/markup-frontstage/package.json:14]: Looks like you're missing a `tslint.json` file or a config within the `package.json`

File: packages/markup-frontstage/package.json (hunk 11 above)
Comment (calebmshafer) [reply packages/markup-frontstage/package.json:52]: I've been meaning to ask about this, any particular reason we shouldn't use them with the `^` in these types of packages?  We don't do it in the iModel.js repository because of the linking directory within the repo but when the packages aren't built within the same repo I don't see an issue with using the `^` to match the peerDep semver.

File: packages/markup-frontstage/package.json (hunk 18 above)
Comment (archana-maharjan) [reply packages/markup-frontstage/package.json]: removed please :)

File: packages/markup-frontstage/package.json (hunk 19 above)
Comment (calebmshafer) [reply packages/markup-frontstage/package.json:13]: All the other packages in the repo have the same issue.  Let's follow up in a separate PR to clean this up.

File: packages/markup-frontstage/package.json (hunk 11 above)
Comment (jason-crow) [reply packages/markup-frontstage/package.json:51]: I think now that we're off of dev builds we should use the ^ and expect versioning is done correctly on the core side. This also provides pressure for that to be the case. The bigger concern would be if there was a backend component this relied on that didn't update at the same time, since it could be problematic for a newer frontend to connect to an older backend, but that isn't applicable in this case

Comment (aruniverse) [reply packages/markup-frontstage/package.json:51]: sounds good to me
